### ISSUE TITLE
Knapsack jest process hangs if any test times out

### ISSUE THREAD

**Issue by samsternatretool** [KnapsackPro/knapsack-pro-js#100]
- [x] I tagged with the correct `@knapsack-pro/PACKAGE` label

In our CI we've seen some hard to debug issues where our normally ~10m test run will hang indefinitely in the middle until we hit the 20m timeout and the runner exits.

After doing a lot of digging, we figured out that what's happening is a single test is hitting the Jest timeout (set at 60s in our codebase) and for some reason this makes the Knapsack process hang forever. What I would expect is for the test to exit and for Knapsack to report it as failed/timed out.

`--forceExit` does not seem to help. The only thing that helps is looking at the logs, figuring out which test did not run, and bumping the timeout for that one test using `jest.setTimeout`

**Comment by ArturT** [comment]
This is how we run Jest:
https://github.com/KnapsackPro/knapsack-pro-js/blob/29ed2ae1dbd28654638e21cd1325c6a2052d9958/packages/jest/src/knapsack-pro-jest.ts#L51

Perhaps we should handle some kind of event when the timeout is exceeded? What else could be causing this issue?

What's your default timeout? Do you configure it somewhere like a jest config file?

**Comment by ArturT** [comment]
Have you tried the following?
https://docs.knapsackpro.com/jest/troubleshooting/#jest-hangs-after-running-tests

* `--detectOpenHandles`
* increase the memory available to Node with [--max_old_space_size](https://nodejs.org/api/cli.html#--max-old-space-sizesize-in-megabytes)

Is the test still hanging?

**Comment by samsternatretool** [comment]
I agree the way you're calling `runCLI` looks right but there must be something missing. The "fix" for us was to call `jest.setTimeout(180)` inside the test suite to give it more time. It was previously set at `jest.setTimeout(120)` but was sometimes exceeding the limit and causing Knapsack to hang.

I have not tried `--detectOpenHandles` as that comes with `runInBand` which we're not currently using (I know it's what Knapsack recommends, but due to the size of our worker machines we get the best results with multiple jest workers per shard).

I imagine you will be able to repro this if you create a `jest` test that has a long `wait()` in it.

**Comment by ArturT** [comment]
I'll try to reproduce it and post info if we come up with any solution.

**Comment by ArturT** [comment]
I can't reproduce the error. I must be missing something.

I configured a sleep to 2000ms and the timeout for the test to 1000ms in [a spec](https://github.com/KnapsackPro/knapsack-pro-js/blob/main/packages/jest-example-test-suite/__tests__/a.test.js).

```typescript
// packages/jest-example-test-suite/__tests__/a.test.js

const aModule = require('../src/a');

const sleep = waitTimeInMs =>
  new Promise(resolve => setTimeout(resolve, waitTimeInMs));

// ...

// configured here
it("sleep", async () => {
  await sleep(2000);
}, 1000);
```

After 1s, the test fails. I'm getting the following error message as expected.

```
... run tests fetched from the Knapsack Pro Queue API

 PASS  __tests__/directory/c.test.js
 PASS  __tests__/b.spec.js
 FAIL  __tests__/a.test.js
  ● sleep

    thrown: "Exceeded timeout of 1000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      14 | });
      15 |
    > 16 | it("sleep", async () => {
         | ^
      17 |   await sleep(2000);
      18 | }, 1000);
      19 |

      at Object.it (__tests__/a.test.js:16:1)

A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.
Test Suites: 1 failed, 2 passed, 3 total
Tests:       1 failed, 7 passed, 8 total
Snapshots:   0 total
Time:        1.927 s, estimated 2 s
Ran all test suites within paths "__tests__/a.test.js", "__tests__/b.spec.js", "__tests__/directory/c.test.js".


... pull more tests from Knapsack Pro Queue API
... exit process with 1 exit code as expected
```

If I increase the sleep to 20s, then I'm getting the error after 1s as expected. No hanging process. 

I use Knapsack Pro with this bin script: [`bin/knapsack_pro_jest`](https://github.com/KnapsackPro/knapsack-pro-js/blob/b5fe743efd4dfcb81706862049785fb0dc7edcdb/packages/jest-example-test-suite/bin/knapsack_pro_jest#L17)

I use Jest 29.7.0. @samsternatretool What's your version of Jest?

Do you see any error message on the output while the CI node hangs? Is it similar to the error I'm getting?

What does your test do? Perhaps it prevents Jest to gracefully handle exit from the test when the timeout is exceeded.

**Comment by samsternatretool** [comment]
@ArturT thanks for trying to repro this! We're using jest `^29.5.0` today, and actually we just ran into this issue again this morning. We were able to narrow down the offending test by turning on Knapsack debug logging to see which tests it was trying to run when it timed out. There's no error message at all, because it just hangs until our CI provider (Buildkite) kills the job. The last log line we see before the timeout is the previous test suite passing:

```
PASS frontend-integration src/store/workflows/workflowsSlice.spec.ts (74.742 s)
# Received cancellation signal, interrupting
Terminated
🚨 Error: The command exited with status 143
```

Our tests are fairly complex (testing React components) so I don't think I'll be able to create a shareable isolated reproduction case, but I'll see if I can identify anything else interesting about them.

**Comment by ArturT** [comment]
I don't know if this helps in case of Jest, but you may take a look at the `timeout` program and use it to send a termination signal to the process before CI kills it so that you capture an output in CI logs. Perhaps Jest would print more info if it receives the `TERM` signal.

<img width="984" alt="image" src="https://github.com/user-attachments/assets/5544d097-7e1d-413f-a4c5-e5503c531393" />

Source: https://docs.knapsackpro.com/ruby/troubleshooting/#why-does-the-ci-node-hang-backtrace-debugging

Ensure you use [the `--verbose` option](https://jestjs.io/docs/cli#--verbose) passed to Knapsack Pro.

```
timeout --signal=TERM 600 npx knapsack-pro-jest --verbose
```

Maybe this way you would be able to narrow down the test case and some backtrace that causes Jest to hang.

You may revisit trying `--detectOpenHandles` too:
https://docs.knapsackpro.com/jest/troubleshooting/#jest-hangs-after-running-tests

Please let me know what your findings are.

**Comment by ArturT** [comment]
I'm closing this as it is a stale issue.

If you still need some help, please reopen it or contact [the support email](https://knapsackpro.com/contact). I'd be happy to help.
